statesync/snapshots.go

```diff
@@ -0,0 +1,262 @@
+package statesync
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	lite "github.com/tendermint/tendermint/lite2"
+	"github.com/tendermint/tendermint/p2p"
+)
+
+const ()
+
+// snapshotHash is a snapshot hash, used for lookups.
+type snapshotHash [sha256.Size]byte
+
+// snapshot contains data about a snapshot.
+type snapshot struct {
+	Height      uint64
+	Format      uint32
+	ChunkHashes [][]byte
+	Metadata    []byte
+
+	trustedAppHash []byte // populated by light client
+}
+
+// Hash generates a snapshot hash, used for lookups. It takes into account not only the height
+// and format, but also the chunk hashes and metadata in case peers have generated snapshots in a
+// non-deterministic manner such that chunks from different peers wouldn't fit together.
+func (s *snapshot) Hash() snapshotHash {
+	// Hash.Write() never returns an error.
+	hasher := sha256.New()
+	hasher.Write([]byte(fmt.Sprintf("%v:%v", s.Height, s.Format)))
+	if s.Metadata != nil {
+		hasher.Write(s.Metadata)
+	}
+	for _, chunkHash := range s.ChunkHashes {
+		hasher.Write(chunkHash)
+	}
+	var hash snapshotHash
+	copy(hash[:], hasher.Sum(nil))
+	return hash
+}
+
+// snapshotPool discovers and aggregates snapshots across peers.
+type snapshotPool struct {
+	lc *lite.Client
+
+	sync.Mutex
+	snapshots     map[snapshotHash]*snapshot
+	snapshotPeers map[snapshotHash]map[p2p.ID]p2p.Peer
+
+	// indexes for fast searches
+	formatIndex map[uint32]map[snapshotHash]bool
+	heightIndex map[uint64]map[snapshotHash]bool
+	peerIndex   map[p2p.ID]map[snapshotHash]bool
+
+	// blacklists for rejected items
+	formatBlacklist   map[uint32]bool
+	heightBlacklist   map[uint64]bool
+	snapshotBlacklist map[snapshotHash]bool
+}
+
+// newSnapshotPool creates a new snapshot pool.
+func newSnapshotPool(lc *lite.Client) *snapshotPool {
+	return &snapshotPool{
+		lc:                lc,
+		snapshots:         make(map[snapshotHash]*snapshot),
+		snapshotPeers:     make(map[snapshotHash]map[p2p.ID]p2p.Peer),
+		formatIndex:       make(map[uint32]map[snapshotHash]bool),
+		heightIndex:       make(map[uint64]map[snapshotHash]bool),
+		peerIndex:         make(map[p2p.ID]map[snapshotHash]bool),
+		formatBlacklist:   make(map[uint32]bool),
+		heightBlacklist:   make(map[uint64]bool),
+		snapshotBlacklist: make(map[snapshotHash]bool),
+	}
+}
+
+// Add adds a snapshot to the pool, unless the peer has already sent recentSnapshots snapshots. It
+// returns true if this was a new, non-blacklisted snapshot. The snapshot height is verified using
+// the light client, and the expected app hash is set for the snapshot.
+func (p *snapshotPool) Add(peer p2p.Peer, snapshot *snapshot) (bool, error) {
+	// FIXME Check if the light client deduplicates concurrent requests for the same height.
```

Comment:
https://github.com/tendermint/tendermint/blob/379848eab7967186e13c7e50b35e512749e8f4f9/lite2/client_test.go#L622-L664

Reply:
Does this mean that the light client is not concurrency-safe _at all_? Or just that it doesn't deduplicate requests?

Reply:
> VerifyHeaderAtHeight is not concurrency-safe at all

not concurrency safe.

in general, we follow Go doc style - it's assumed func/object is not safe for concurrent access unless stated (via comment) otherwise.

Reply:
Ok, thanks. Would be nice if it was. :) But I'll wrap it in a mutex for now.